### PR TITLE
Revert hvd.DistributedOptimizer to allreduce in compute_gradients(), require graph mode

### PR DESCRIPTION
**Pull request by alsrgv**
Since landing #1311 and #1316 we have identified few issues with that proposal:
1. Inconsistency between PyTorch and TensorFlow in terms of handling of gradient clipping (see https://github.com/horovod/horovod/issues/1346).
2. Incompatibility with loss scaling in NVIDIA AMP and native TensorFlow loss scaling optimizer.

The idea is to revert the change and instead require graph mode in TF2-Keras.  Since this is another behavior change, we'll need to release 0.18.0.

cc @DEKHTIARJonathan

### CODE REVIEW

File: horovod/_keras/__init__.py [hunk 1]
```diff
@@ -60,22 +40,42 @@ def get_gradients(self, loss, params):
...
+                                                     compression=self._compression)
+                            averaged_gradients.append(avg_grad)
+                        else:
+                            averaged_gradients.append(None)
+                    return averaged_gradients
             else:
                 return gradients
 
+        def apply_gradients(self, *args, **kwargs):
+            self._fail_if_executing_eagerly()
```
Comment (romerojosh) [inline horovod/_keras/__init__.py]: Is it possible to just have the eager failure occur during `__init__` instead of in both `apply_gradients` and `compute_gradients`?

File: horovod/tensorflow/__init__.py [hunk 2]
```diff
@@ -243,23 +242,27 @@ def __init__(self, optimizer, name=None, use_locking=False, device_dense='',
...
 
-            See Optimizer.apply_gradients() for more info.
+            See Optimizer.compute_gradients() for more info.
 
-            In DistributedOptimizer, apply_gradients() is overriden to also
-            allreduce the gradients before applying them.
+            In DistributedOptimizer, compute_gradients() is overriden to also
+            allreduce the gradients before returning them.
             """
+            self._fail_if_executing_eagerly()
```
Comment (romerojosh) [inline horovod/tensorflow/__init__.py]: Same here. Is it possible to just have the eager failure occur during `__init__` instead of in both `apply_gradients` and `compute_gradients`?

File: horovod/_keras/__init__.py [hunk 3]
```diff
@@ -60,22 +40,42 @@ def get_gradients(self, loss, params):
...
+            return super(self.__class__, self).apply_gradients(*args, **kwargs)
+
+        def _fail_if_executing_eagerly(self):
+            if hvd._executing_eagerly():
```
Comment (tgaddair) [inline horovod/_keras/__init__.py]: Does this behavior differ between TF 2.0 and older versions?

File: horovod/_keras/__init__.py (hunk 1 above)
Comment (tgaddair) [reply horovod/_keras/__init__.py]: My understanding is that at the time the optimizer is created it's fine to be running eagerly, only during fit you need to be in graph mode.

Comment (alsrgv) [reply horovod/_keras/__init__.py]: Yeah, that's exactly right.

File: horovod/_keras/__init__.py (hunk 3 above)
Comment (alsrgv) [reply horovod/_keras/__init__.py]: In order versions you need to opt in to eager mode. @romerojosh has a good point that eager mode is slower than graph mode, so if somebody is throwing a lot of resources at a problem, they should switch to graph mode first.

File: horovod/_keras/__init__.py (hunk 1 above)
Comment (romerojosh) [reply horovod/_keras/__init__.py]: I figured that might be the case but wanted to check. Thanks!

**Comment by alsrgv** [comment]
@romerojosh, @tgaddair, I had to make a fix. `run_eagerly` was insufficient because it was off by default.  TF 2.0 uses `tf.function` to run `fit()` which runs the actual training in graph mode, but it's still using GradientTape to compute the gradient.  I've modified the example to disable that behavior and the optimizer to detect if get_gradient() was never called.
